### ISSUE TITLE
Derive and ship the L* form of the lossy objective (follow-up to #1011)

### ISSUE THREAD

**Issue by danieledler** [mapequation/infomap#1024]
Follow-up to **#1011**, which rejects `--lossy --non-redundant` rather than deriving it. The derivation
itself is **done and verified numerically**; what stopped it shipping is the reporting layer, and that is
what this issue is for. Written up as **F40** in `columnar_wip/columnar-rethink-notes.md` on
`columnar-hierarchical-core`.

Everything below needs `FEATURES=lossy-map-equation`. Measurements are on a clean
`columnar-hierarchical-core` d88f1c77 release build,
`FEATURES="lossy-map-equation regularized-multilayer" OPENMP=0`, md5 `b34987647a66051a0bf038d6a43e736a`
— i.e. **before** #1011's rejection lands, which is the last point at which the combination runs.

## The defect, in one measurement

The noise credit is objective-independent, which a correctly derived credit cannot be. Fixed partition,
`test/fixtures/networks/lossy_benchmark.net` + `test/fixtures/clusters/lossy_benchmark.clu
--no-infomap`:

| `--lambda` | `-C --lossy` (L) | `--non-redundant --lossy` (L\*) | L − L\* |
|---:|--:|--:|--:|
| 1 | 2.4099867595799425 | 2.3521420558461954 | 0.05784470373374706 |
| 1.5 | 2.653992887305974 | 2.596148183572227 | 0.05784470373374706 |
| 2 | 2.7309159642290513 | 2.673071260495304 | 0.05784470373374706 |
| 2.5 | 2.807839041152128 | 2.749994337418381 | 0.05784470373374706 |
| 3 | 2.818018368324836 | 2.760173664591089 | 0.05784470373374706 |
| 4 | 2.818018368324836 | 2.760173664591089 | 0.05784470373374706 |
| 5 | 2.818018368324836 | 2.760173664591089 | 0.05784470373374706 |
| *(no `--lossy`)* | 2.818018368324836 | 2.760173664591089 | 0.05784470373374706 |

The credit subtracted from L\* is bit-for-bit the credit subtracted from L, at every λ including the ones
where the gate is shut. Only a hardcoded coefficient 1 produces that. `LossyCorrection` is a
*substituting* correction, so it inherits the surrounding objective's rate — the same argument that
#1009 established for `MemCorrection`.

## The derived form

```
correction_m = -max(0, rate_m * l_m - lambda * H_m)
```

with `rate_m = nrLeafCodebookRate(...)` — the same per-module rate `ColumnarTwoLevel::leafCodebookRates()`
(`src/core/ColumnarObjectiveScore.cpp:257`) already hands `MemCorrection`. The multiplier sits **inside**
the `max(0, .)` and multiplies `l_m` only, so it changes the **gate**, not merely the accounting. That is
what makes this different from #1010, where the rate only re-weighted a term that was already being
charged.

Today's `LossyCorrection::moduleCost` (`src/core/ColumnarCorrections.cpp:637`) has no rate at all:

```cpp
  return std::max(0.0, (plogp(flow) - flowLogFlow) - m_lambda * entropy);
```

### Verified by building it

A throwaway build of fix step 1 only — `hierarchicalCorrection` fetching `core.leafCodebookRates()` and
summing `max(0, rates[m] * (plogp(F[m]) - flf[m]) - lambda * H[m])` when non-empty (never committed,
md5 `7a136f8f9895a51730874bcd66d09365`):

- λ = 1, fixed partition: **2.3383174785435856** (vs 2.3521420558461954 unpatched)
- `-C --lossy` unchanged at **2.4099867595799425** — the `rates.empty()` branch really is inert, so the
  base objective stays bit-identical.

## Why it is not shippable as a one-line change

Fix step 1 alone produces a self-contradicting result, because the reporting layer is base-flavoured.
Half-fix build, `--non-redundant --lossy --lambda 2.58 -N5 --seed 123`:

```
# codelength 2.75792 bits
# lossy lambda 2.58 J 2.75792 rate 2.81802 distortion 0 one-level lossless 3.79988
# noise modules 0 of 3: (none)
```

against plain L\* on the same input of **2.760173664591089**. A `J` strictly *below* the lossless L\*,
with zero distortion and an empty noise set. The unpatched binary prints the consistent-looking
`J 2.76017` for the same run.

Two reporters have to move with the correction:

1. **`InfomapBase::noiseTopModules()` (`src/core/InfomapBase.cpp:2270`)** gates on the base rule and
   never sees a rate:

   ```cpp
     const double loss = infomath::plogp(module.data.flow) - module.lossyFlowLogFlow;
     if (loss > lossyLambda * module.lossyEntropy + 1e-12)
   ```

   It must use the same gate, which means the columnar trial stashing the gated module ids rather than
   this function recomputing L\* boundary rates on the object-oriented tree.

2. **`getLossyRate()` / `getLossyDistortion()` (`src/core/InfomapBase.h:184-185`, forwarded to the
   object-oriented `LossyMapEquation` via `src/core/InfomapOptimizer.h:188-196`)** must become the L\*
   decomposition, or `J == rate + lambda * distortion` stays broken.

## Definition of done

The invariant already has a test; it just has no L\* arm (see the companion test-coverage issue). Add a
third element to the loop in `test/cpp/test_lossy.cpp:113`:

```cpp
  for (const auto* lambdaFlag : { "--lossy --lambda 1.5", "--lossy --lambda 2",
                                  "--non-redundant --lossy --lambda 1.5" }) {
```

It fails today — `J 2.59615` against `rate + lambda * distortion = 2.42322 + 1.5 * 0.153846 = 2.653989`
— and can only pass once all three pieces land. Note it **cannot be added before #1011's rejection is
lifted**, since the combination throws in the meantime; that ordering is part of this issue's scope.

## Notes

- The modelling assumption behind the derivation — that a noise module under L\* is the same L\* module
  term with its node set collapsed to one codeword (enter codebook -> 0 bits, within-codebook usage and
  `lambda * H` unchanged) — is stated, not proved. The numeric agreement between a from-scratch model
  and the patched build is a coding check, not corroboration of the model. F40 says so.
- Silent counterexamples exist and the fix must keep them silent. Two disjoint 4-cliques partitioned
  into their components have `rate == 1` in every module, and score identically on both arms:

  ```
  $ printf '*Edges\n1 2 1\n1 3 1\n1 4 1\n2 3 1\n2 4 1\n3 4 1\n5 6 1\n5 7 1\n5 8 1\n6 7 1\n6 8 1\n7 8 1\n' > $D/cliques.net
  $ printf '# m\n1 1\n2 1\n3 1\n4 1\n5 2\n6 2\n7 2\n8 2\n' > $D/cliques.clu
  $ ./Infomap $D/cliques.net $OUT -c $D/cliques.clu --no-infomap --silent <arm> --summary-json -
      -C --lossy               1.584962500721156
      --non-redundant --lossy  1.584962500721156
  ```

  And from `λ ≥ 3` in the table above the gate is shut, so each arm returns exactly its own
  no-`--lossy` value. Neither case may move.
- `--non-redundant` forces `columnarSearch`, and `m_nonRedundant` is read in exactly two places in
  `src/` (`src/core/ColumnarObjectiveScore.cpp:253` and `:270`), so the blast radius of the correction
  itself is small. It is the two reporters above that reach outside the columnar core.

`Follow-up to #1011. Related to #1009, #1010.`
